Registration/RegisterWithAzure.psm1

```diff
@@ -1649,56 +1640,48 @@ function Get-AzToken
 
-        if (-not $FromCache)
+        #
+        # Resolve token cache data
+        #
+        $accounts = $null
+        if ((Get-Module -Name "Az.Accounts").Version -le [Version]"2.0.1")
         {
-            $token = [Microsoft.Azure.Commands.Common.Authentication.AzureSession]::Instance.AuthenticationFactory.Authenticate(
-                ($account            = $Context.Account),
-                ($environment        = $Context.Environment),
-                ($tenant             = $TenantId),
-                ($password           = $null),
-                ($promptBehavior     = 'Never'),
-                ($promptAction       = $null),
-                ($tokenCache         = $null),
-                ($resourceIdEndpoint = $Resource))
-
-            return [pscustomobject]@{ AccessToken = ConvertTo-SecureString $token.AccessToken -AsPlainText -Force } |
-                Add-Member -MemberType ScriptMethod -Name 'GetAccessToken' -Value { return [System.Net.NetworkCredential]::new('$tokenType', $this.AccessToken).Password } -PassThru
+            [Microsoft.Azure.Commands.Common.Authentication.Authentication.Clients.AuthenticationClientFactory]$authenticationClientFactory = $null
+            if (-not ([Microsoft.Azure.Commands.Common.Authentication.AzureSession]::Instance.TryGetComponent(
+                [Microsoft.Azure.Commands.Common.Authentication.Authentication.Clients.AuthenticationClientFactory]::AuthenticationClientFactoryKey,
+                [ref]$authenticationClientFactory)))
+            {
+                $m = 'Please ensure you have authenticated with Az Accounts module!'
+                $m += ' Unable to resolve authentication client factory from Az Accounts module runtime'
+                $m += ' ([Microsoft.Azure.Commands.Common.Authentication.Authentication.Clients.AuthenticationClientFactory])'
+                Write-Error $m
+                return
+            }
+
+            $client = $authenticationClientFactory.CreatePublicClient(
+                ($clientId='1950a258-227b-4e31-a9cf-717495945fc2'),
+                ($TenantId),
+                ($authority="$($Context.Environment.ActiveDirectoryAuthority.TrimEnd('/'))/$TenantId"),
+                ($redirectUri='urn:ietf:wg:oauth:2.0:oob'),
+                ($useAdfs=$Context.Environment.ActiveDirectoryAuthority -like '*/adfs*'))
+
+            $authenticationClientFactory.RegisterCache($client)
+            $accounts = $client.GetAccountsAsync().ConfigureAwait($true).GetAwaiter().GetResult()
+            $bytes = ([Microsoft.Identity.Client.ITokenCacheSerializer]$client.UserTokenCache).SerializeMsalV3()
         }
         else
         {
-            Write-Warning "Attempting to find a refresh token and an access token from the existing token cache data..."
+            $provider = [Microsoft.Azure.Commands.Common.Authentication.SharedTokenCacheProvider]::new()
```

Comment:
Do we need a similar change in identity scripts?

Reply:
yes - https://github.com/Azure/AzureStack-Tools/blob/46071b431910a4157ce9ac35697876bbf3676c47/Identity/AzureStack.Identity.Common.psm1#L111

we can't remove the accesstoken part in the identity script as it uses access token too

Reply:
I think the preference would be to use the new Get-AzsToken command (from the new az module) separately for access tokens, and only use the export-refreshtoken function for the refresh token. In the identity script, perhaps we can keep the "aggregate" function on the top-level which can turn-around and call the other two functions ionternally, along with whatever fallback for backwards compatibility that is required?

Reply:
Your new suggested approach can get the access token from cache too. I just updated the function in the identity common script. can you please take a look at this commit https://github.com/Azure/AzureStack-Tools/pull/600/commits/1696b224d484dd6c0451e8ba7a7f1a7ff9be2620